typing/typecore.ml

```diff
@@ -2948,59 +2947,107 @@ let rec approx_type env sty =
         let tyl = List.map (approx_type env) ctl in
         newconstr path tyl
       end
-  | Ptyp_poly (_, sty) ->
+  | Ptyp_poly _ ->
       approx_type env sty
   | _ -> newvar ()
 
-let rec type_approx env sexp =
-  match sexp.pexp_desc with
-    Pexp_let (_, _, e) -> type_approx env e
-  | Pexp_fun (p, _, spat, e) ->
-      let marg =
-        if has_local_attr_pat spat then Alloc_mode.local
-        else Alloc_mode.newvar ()
+let type_pattern_approx env spat ty_expected =
+  match spat.ppat_desc with
+  | Ppat_constraint(_, ({ptyp_desc=Ptyp_poly _} as sty)) ->
+      let arg_type_mode =
+        if has_local_attr_pat spat then Alloc_mode.Local
+        else Alloc_mode.Global
       in
-      let mret = Alloc_mode.newvar () in
-      let ty = if is_optional p then type_option (newvar ()) else newvar () in
-      let ret = type_approx env e in
-      newty (Tarrow((p,marg,mret), ty, ret, Cok))
+      let ty_pat =
+        Typetexp.transl_simple_type env false arg_type_mode sty
+      in
+      begin try unify env ty_pat.ctyp_type ty_expected with Unify trace ->
+        raise(Error(spat.ppat_loc, env, Pattern_type_clash(trace, None)))
+      end;
+  | _ -> ()
+
+let rec type_function_approx env loc label spato sexp in_function ty_expected =
+  let has_local, has_poly =
+    match spato with
+    | None -> false, false
+    | Some spat ->
+        let has_local = has_local_attr_pat spat in
+        let has_poly = has_poly_constraint spat in
+        if has_poly && is_optional label then
+          raise(Error(spat.ppat_loc, env, Optional_poly_param));
+        has_local, has_poly
+  in
+  let loc_fun, ty_fun =
+    match in_function with
+    | Some (loc, ty) -> loc, ty
+    | None -> loc, ty_expected
+  in
+  let (arg_mode, ty_arg, _, ty_res) =
+    try filter_arrow env ty_expected label ~force_poly:(not has_poly)
+    with Unify _ ->
+      match expand_head env ty_expected with
+      | {desc = Tarrow _} as ty ->
+          raise(Error(loc, env, Abstract_wrong_label(label, ty, None)))
+      | _ ->
+          raise(Error(loc_fun, env,
+            Too_many_arguments (in_function <> None, ty_fun, None)))
+  in
+  if has_local then
+    eqmode ~loc ~env arg_mode Alloc_mode.local
+      (Param_mode_mismatch ty_expected);
+  if has_poly then begin
+    match spato with
+    | None -> ()
+    | Some spat -> type_pattern_approx env spat ty_arg
```

Comment:
Can this be done in the earlier match against `spato`? Doing so would avoid the impossible `None` case just above.

Reply:
I think it would be difficult to do so. This call needs `ty_arg`, which comes from the call to `filter_arrow`, which requires `has_poly`.